File: backend/app/api/v1/optimizations.py
"""
Optimization API Routes.
"""
import logging
import uuid

from fastapi import APIRouter, Depends, HTTPException, Query, Request, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.deps import get_current_user
from app.core.rate_limit import limiter
from app.database.session import get_async_session
from app.models.user import User
from app.repositories import optimization_repo, resume_match_repo, resume_repo
from app.schemas.optimization import (
    OptimizationListResponse,
    OptimizationResponse,
    OptimizationRunResponse,
    PromptResponse,
    SaveOptimizationsRequest,
    UpdateOptimizationStatusRequest,
)
from app.services.ats_service import ats_service
from app.services.optimization_service import GEMINI_RESPONSE_SCHEMA, optimization_generator_service

logger = logging.getLogger(__name__)

router = APIRouter()


# ---------------------------------------------------------------------------
# Helper: fetch ATS + JD context for a resume
# ---------------------------------------------------------------------------

async def _fetch_analysis_context(
    db: AsyncSession,
    resume_id: uuid.UUID,
    user_id: uuid.UUID,
    job_id: uuid.UUID | None,
):
    """
    Shared helper that loads ATS analysis and JD match data for a resume.
    Used by both the platform-key generation endpoint and the BYOK prompt endpoint.
    """
    ats_analysis = None
    jd_match = None

    # Fetch ATS analysis
    try:
        analysis_record = await ats_service.get_analysis(db, resume_id, user_id)
        if analysis_record:
            ats_analysis = {"suggestions": getattr(analysis_record, "missing_formatting", [])}
    except HTTPException:
        pass

    # Fetch JD match if job_id provided
    if job_id:
        match_record = await resume_match_repo.get_match(db, resume_id, job_id)
        if match_record:
            jd_match = {
                "missing_skills": match_record.missing_skills,
                "missing_keywords": match_record.missing_keywords
            }

    return ats_analysis, jd_match


# ---------------------------------------------------------------------------
# POST /resumes/{resume_id}/optimize  — Platform key generation (existing)
# ---------------------------------------------------------------------------

@router.post(
    "/resumes/{resume_id}/optimize",
    response_model=OptimizationRunResponse,
    status_code=status.HTTP_201_CREATED,
    tags=["Resume Optimizations"]
)
@limiter.limit("5/minute")
async def generate_resume_optimizations(
    request: Request,
    resume_id: uuid.UUID,
    job_id: uuid.UUID | None = Query(None, description="Optional job description ID to optimize against"),
    db: AsyncSession = Depends(get_async_session),
    current_user: User = Depends(get_current_user),
):
    """
    Generates new optimization suggestions for a resume using the platform's
    Gemini API key.
    """
    resume = await resume_repo.get_by_id(db, resume_id, current_user.id)
    if not resume:
        raise HTTPException(status_code=404, detail="Resume not found")

    ats_analysis, jd_match = await _fetch_analysis_context(db, resume_id, current_user.id, job_id)

    try:
        suggestions_raw = await optimization_generator_service.generate_suggestions(
            resume_text=resume.raw_text,
            ats_analysis=ats_analysis,
            jd_match_results=jd_match
        )
    except Exception as e:
        logger.error(f"Error generating optimizations: {e}", exc_info=True)
        error_msg = str(e)
        if "503" in error_msg or "temporary" in error_msg.lower():
            raise HTTPException(
                status_code=status.HTTP_503_SERVICE_UNAVAILABLE,
                detail=(
                    "CareerPilot AI quota is currently exhausted. "
                    "You can try again later or use your own Gemini API key."
                )
            )
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, 
            detail=f"Failed to generate optimizations: {error_msg}"
        )

    run = await optimization_repo.create_many(
        session=db,
        resume_id=resume_id,
        job_id=job_id,
        model_name=optimization_generator_service.model_name,
        prompt_version=optimization_generator_service.prompt_version,
        suggestions=suggestions_raw
    )
    
    return run


# ---------------------------------------------------------------------------
# GET /resumes/{resume_id}/optimize/prompt  — BYOK prompt endpoint (new)
# ---------------------------------------------------------------------------

@router.get(
    "/resumes/{resume_id}/optimize/prompt",
    response_model=PromptResponse,
    tags=["Resume Optimizations"]
)
async def get_optimization_prompt(
    resume_id: uuid.UUID,
    job_id: uuid.UUID | None = Query(None, description="Optional job description ID to optimize against"),
    db: AsyncSession = Depends(get_async_session),
    current_user: User = Depends(get_current_user),
):
    """
    Returns the assembled AI prompt and JSON schema so the frontend can call
    the Gemini API directly using the user's own API key.

    The user's API key never touches this server.
    """
    resume = await resume_repo.get_by_id(db, resume_id, current_user.id)
    if not resume:
        raise HTTPException(status_code=404, detail="Resume not found")
    if not resume.raw_text:
        raise HTTPException(status_code=400, detail="Resume text not yet parsed or is empty.")

    ats_analysis, jd_match = await _fetch_analysis_context(db, resume_id, current_user.id, job_id)

    prompt = optimization_generator_service.build_prompt(
        resume_text=resume.raw_text,
        ats_analysis=ats_analysis,
        jd_match_results=jd_match
    )

    return PromptResponse(
        prompt=prompt,
        model_name=optimization_generator_service.model_name,
        response_schema=GEMINI_RESPONSE_SCHEMA,
    )


# ---------------------------------------------------------------------------
# POST /resumes/{resume_id}/optimize/save  — BYOK save endpoint (new)
# ---------------------------------------------------------------------------

@router.post(
    "/resumes/{resume_id}/optimize/save",
    response_model=OptimizationRunResponse,
    status_code=status.HTTP_201_CREATED,
    tags=["Resume Optimizations"]
)
async def save_byok_optimizations(
    resume_id: uuid.UUID,
    body: SaveOptimizationsRequest,
    job_id: uuid.UUID | None = Query(None, description="Optional job description ID"),
    db: AsyncSession = Depends(get_async_session),
    current_user: User = Depends(get_current_user),
):
    """
    Persists optimization suggestions that were generated client-side
    (via the user's own Gemini API key) into the database.

    The user's API key is never sent to this endpoint.
    """
    resume = await resume_repo.get_by_id(db, resume_id, current_user.id)
    if not resume:
        raise HTTPException(status_code=404, detail="Resume not found")

    if not body.suggestions:
        raise HTTPException(status_code=400, detail="No suggestions provided.")

    run = await optimization_repo.create_many(
        session=db,
        resume_id=resume_id,
        job_id=job_id,
        model_name=body.model_name,
        prompt_version=body.prompt_version,
        suggestions=body.suggestions,
    )

    return run


# ---------------------------------------------------------------------------
# GET /resumes/{resume_id}/optimizations  — List (existing)
# ---------------------------------------------------------------------------

@router.get(
    "/resumes/{resume_id}/optimizations",
    response_model=OptimizationListResponse,
    tags=["Resume Optimizations"]
)
async def list_resume_optimizations(
    resume_id: uuid.UUID,
    db: AsyncSession = Depends(get_async_session),
    current_user: User = Depends(get_current_user),
):
    """
    Lists all optimizations generated for a resume.
    """
    resume = await resume_repo.get_by_id(db, resume_id, current_user.id)
    if not resume:
        raise HTTPException(status_code=404, detail="Resume not found")
        
    optimizations = await optimization_repo.get_for_resume(db, resume_id)
    return OptimizationListResponse(optimizations=list(optimizations), total_count=len(optimizations))


# ---------------------------------------------------------------------------
# PATCH /optimizations/{optimization_id}/status  — Update status (existing)
# ---------------------------------------------------------------------------

@router.patch(
    "/optimizations/{optimization_id}/status",
    response_model=OptimizationResponse,
    tags=["Resume Optimizations"]
)
async def update_optimization_status(
    optimization_id: uuid.UUID,
    request: UpdateOptimizationStatusRequest,
    db: AsyncSession = Depends(get_async_session),
    current_user: User = Depends(get_current_user),
):
    """
    Updates the status (e.g. ACCEPTED, REJECTED) of a specific optimization suggestion.
    """
    opt = await optimization_repo.update_status(
        session=db, 
        optimization_id=optimization_id, 
        status=request.status,
        user_id=current_user.id
    )
    if not opt:
        raise HTTPException(status_code=404, detail="Optimization not found")
    return opt

File: frontend/src/lib/api/optimizations.ts
// Optimizations API client
import { getToken } from "../tokenStore";

const API_BASE = import.meta.env.VITE_API_URL ?? "http://localhost:8000/api/v1";

export type OptimizationStatus = "PENDING" | "ACCEPTED" | "REJECTED";
export type OptimizationType =
  | "KEYWORD"
  | "BULLET_REWRITE"
  | "SKILL_ADDITION"
  | "SUMMARY_IMPROVEMENT"
  | "ATS_FIX";

export interface OptimizationResponse {
  id: string;
  run_id: string;
  optimization_type: OptimizationType;
  section: string;
  original_text: string;
  suggested_text: string;
  reasoning: string;
  status: OptimizationStatus;
  created_at: string;
  updated_at: string;
}

export interface OptimizationRunResponse {
  id: string;
  resume_id: string;
  job_description_id: string | null;
  model_name: string;
  prompt_version: string;
  created_at: string;
  updated_at: string;
  suggestions: OptimizationResponse[];
}

export interface OptimizationListResponse {
  optimizations: OptimizationResponse[];
  total_count: number;
}

export interface PromptResponse {
  prompt: string;
  model_name: string;
  response_schema: Record<string, unknown>;
}

function authHeaders(): HeadersInit {
  const token = getToken();
  return token ? { Authorization: `Bearer ${token}` } : {};
}

async function handleResponse<T>(res: Response): Promise<T> {
  if (!res.ok) {
    const err = await res.json().catch(() => ({}));
    throw new Error(err.detail ?? `Request failed with status ${res.status}`);
  }
  return res.json() as Promise<T>;
}

export const optimizationsApi = {
  /** POST /resumes/{resumeId}/optimize — uses platform key */
  async generate(
    resumeId: string,
    jobId?: string,
  ): Promise<OptimizationRunResponse> {
    const url = jobId
      ? `${API_BASE}/resumes/${resumeId}/optimize?job_id=${jobId}`
      : `${API_BASE}/resumes/${resumeId}/optimize`;

    const res = await fetch(url, {
      method: "POST",
      headers: authHeaders(),
    });
    return handleResponse<OptimizationRunResponse>(res);
  },

  /** GET /resumes/{resumeId}/optimize/prompt — fetch prompt for BYOK */
  async getPrompt(resumeId: string, jobId?: string): Promise<PromptResponse> {
    const url = jobId
      ? `${API_BASE}/resumes/${resumeId}/optimize/prompt?job_id=${jobId}`
      : `${API_BASE}/resumes/${resumeId}/optimize/prompt`;

    const res = await fetch(url, {
      headers: authHeaders(),
    });
    return handleResponse<PromptResponse>(res);
  },

  /** POST /resumes/{resumeId}/optimize/save — persist BYOK results */
  async saveByok(
    resumeId: string,
    suggestions: Record<string, unknown>[],
    jobId?: string,
  ): Promise<OptimizationRunResponse> {
    const url = jobId
      ? `${API_BASE}/resumes/${resumeId}/optimize/save?job_id=${jobId}`
      : `${API_BASE}/resumes/${resumeId}/optimize/save`;

    const res = await fetch(url, {
      method: "POST",
      headers: {
        ...authHeaders(),
        "Content-Type": "application/json",
      },
      body: JSON.stringify({
        model_name: "gemini-2.5-flash",
        prompt_version: "v1.0-byok",
        suggestions,
      }),
    });
    return handleResponse<OptimizationRunResponse>(res);
  },

  /** GET /resumes/{resumeId}/optimizations */
  async listForResume(resumeId: string): Promise<OptimizationListResponse> {
    const res = await fetch(`${API_BASE}/resumes/${resumeId}/optimizations`, {
      headers: authHeaders(),
    });
    return handleResponse<OptimizationListResponse>(res);
  },

  /** PATCH /optimizations/{id}/status */
  async updateStatus(
    optimizationId: string,
    status: OptimizationStatus,
  ): Promise<OptimizationResponse> {
    const res = await fetch(
      `${API_BASE}/optimizations/${optimizationId}/status`,
      {
        method: "PATCH",
        headers: {
          ...authHeaders(),
          "Content-Type": "application/json",
        },
        body: JSON.stringify({ status }),
      },
    );
    return handleResponse<OptimizationResponse>(res);
  },
};
